lib/kdtree.h

```diff
@@ -219,6 +221,36 @@ class FlatNode {
         return static_cast<uint32_t>(data_) >> 2;
     }
 
+    // Since right nodes are create after parent nodes, we need a setter to
+    // update the index of the right node after creation.
+    void set_right(uint32_t value) {
+        data_ = (data_ & 0x0000) |
+                static_cast<uint64_t>(value << 2 |
+                                      static_cast<uint32_t>(split_axis()));
+    }
+
+    // leaf node attributes
+
+    uint32_t first_triangle_id() const {
+        assert(is_leaf());
+        return static_cast<uint32_t>(data_ >> 32);
+    }
+
+    uint32_t second_triangle_id() const {
```

Comment:
We have 8 bytes. For an triangle index (or id), we use 4. It is enough since then we can have 4.294.967.296 triangles indexed. Actually, we don't have the full 4 bytes, but only 30 bits, since we need the first two bits for storing the type of the node: leaf, or inner with axis X, Y, or Z. So we can index only 1.073.741.824 many triangles, which is still enough in my opinion. That's we don't want to waste space, and store directly two triangles in a leaf node. If the leaf node is the last before sentinel, it may happen that we have only one triangle left. Then the second triangle id will be invalid.

This all follows from the description of the memory layout of FlatNode.